### ISSUE TITLE
Mismatched types [E0308] error in IDE: false positives & false negatives

### ISSUE THREAD

**Issue by tongohiti** [intellij-rust/intellij-rust#3514]
## Environment

* **Intellij-Rust plugin version:** v0.2.94.2569-183-nightly
* **Rust toolchain version:** rust-1.32.0-x86_64-pc-windows-gnu
* **IDE name and version:** CLion 2018.3.4
* **Operating system:** Windows 7 x64

## Problem description
IDE sometimes detects false positives on "mismatched types" error, and sometimes false negatives.
Here are the examples of both.

### This code wouldn't compile, but IDE sees no problem
```rust
fn get_byte_slice() -> Result<&'static [u8], ()> {
    Ok(b"")
}

/*
  This example wouldn't compile due to the following two errors:

error[E0308]: mismatched types
  --> test1.rs:31:27
   |
31 |             if !(*left == *right) {
   |                           ^^^^^^ expected slice, found array of 3 elements
   |
   = note: expected type `std::result::Result<&[u8], ()>`
              found type `std::result::Result<&[u8; 3], _>`


error[E0308]: mismatched types
  --> test1.rs:38:5
   |
38 |     assert_eq!(get_byte_slice(), Ok(b"lol"));
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected slice, found array of 3 elements
   |
   = note: expected type `std::result::Result<&[u8], ()>`
              found type `std::result::Result<&[u8; 3], _>`
*/
fn main() {
    // The following snippet mimicks expansion of assert_eq!() macro -- IDEA shows no errors
    match (&get_byte_slice(), &Ok(b"lol")) {
        (left, right) => {
            if !(*left == *right) {
                unimplemented!()
            }
        }
    }

    // This is the inspiration for the above -- again, IDEA wouldn't complain
    assert_eq!(get_byte_slice(), Ok(b"lol"));
}
```

### This code compiles just fine, but IDE complains

```rust
pub enum LolType {
    Value(Vec<&'static[u8]>)
}

fn main() {
    // The following code compiles just fine, but IDEA complains about mismatched types
    let _x = LolType::Value(vec![b"lol"]);
}
```
![test2](https://user-images.githubusercontent.com/15714232/54087624-85a0e380-434c-11e9-8e96-7bb9b3c9add4.png)

**Comment by murtyjones** [comment]
I get this as well. Compiles fine but the IDE gives a false E0308 error.

**Intellij-Rust plugin version:** v0.2.95.2120-183
**Rust toolchain version:** 1.35.0
**IDE name and version:** PHPStorm 2019.1
**Operating system:** macOS

<img width="898" alt="Screen Shot 2019-04-03 at 2 17 33 PM" src="https://user-images.githubusercontent.com/708562/55502776-56178b00-561b-11e9-9e00-9806ba458075.png">

**Comment by elichai** [comment]
The exact same thing happens with Box's:
```
fn main() {
    let mut list: Vec<Box<[u8]>> = Vec::new();
    let test = [1u8; 32];
    list.push(Box::new(test));
}
```

```
mismatched types [E0308] expected `[u8]`, found `[u8; 32]` less... (Ctrl+1) 
Inspection info:The Rust type checker.
```

**Comment by Maksim20023** [comment]
Hi! I’m closing this ticket because the issue does not reproduce in the current environment (CLion 2023.2 Build #CL-232.8660.186, Rust plugin 0.4.199.5415-232). If this issue still reproduces for you, please reopen the ticket.
